### DISCUSSION TITLE
Relation field not expanded when using pocketbase dao

### DISCUSSION THREAD

**Discussion by Clovis-krz** [pocketbase/pocketbase#2904] [Q&A] [answered]
Expanding works perfectly with the dart sdk. but when i try to access the db in the backend with dao i cannot make a research with expanded fields neither expanding the result.
Example :
```
records, err := app.Dao().FindRecordsByExpr("articles",
			dbx.NewExp("author.name = {:author_name}", dbx.Params{"author_name": "Jack"}),
		).ExpandRecords()
if err != nil {
   print(err)
}
print(records)
```
Considering author to be a relation to another collection containing a name column.
This will tell me that the length of records is 0 because even if i have articles with author named Jack linked.
I didn't find any information about expanding relations in here : https://pocketbase.io/docs/record-methods/

**Comment by ganigeorgiev** [comment]
For expanding relations programmatically, you can check the example in https://github.com/pocketbase/pocketbase/discussions/2901#discussioncomment-6425293.

Searching in nested relation fields is a little more complicated because it requires manually binding the relation joins. With the upcoming v0.17.0 release there will be 2 new helpers [`Dao.FindRecordsByFilter()`](https://github.com/pocketbase/pocketbase/blob/7bb33d4453453731cd01d6d810105518cd405593/daos/record.go#L239) and [`Dao.FindFirstRecordByFilter`](https://github.com/pocketbase/pocketbase/blob/7bb33d4453453731cd01d6d810105518cd405593/daos/record.go#L303) that can help with this.

If the above doesn't help, please elaborate and provide a little more context about your use case and what are you trying to do.

**Reply by Clovis-krz** [reply]
Thanks a lot for your reply! [Dao.FindRecordsByFilter()](https://github.com/pocketbase/pocketbase/blob/7bb33d4453453731cd01d6d810105518cd405593/daos/record.go#L239) is exactly what i was looking for and I am looking forward to it. When do you think 0.17.0 will be out ?

**Reply by ganigeorgiev** [reply]
Sorry, there are no hard ETAs, if nothing unexpected comes it should be released within the next couple weeks (_I'm still doing some experiments with the hooks and don't want to rush it since after v0.17.0 we'll have also js hooks support and I want to avoid introducing breaking changes in the future as much as possible_).

I have no context about your use case but keep in mind that the above helpers are only for retrieving records with a filter string. The relations still will not be expanded. If you want expand, you have to provide a little more information which relations you want to expand (single and/or multiple), whether access checks need to be performed for the expansion, how do you plan to interact with the expanded records (eg. does `map[string]any` satisfy your use case?), etc.

I asked for more details about the use case, because if you only want to return "enriched" records response from a custom route with the appropiate access checks, you can make use of [`apis.EnrichRecord()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.8/apis#EnrichRecord) and [`apis.EnrichRecords()`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.16.8/apis#EnrichRecords) helpers (_both internally calls `Dao.ExpandRecord()` by providing a special filter function that checks the View rule of the related collection(s) against the request data_).

**Reply by Clovis-krz** [reply]
I plan to use it in a cron task, not as a custom route. I need full access to the collections with no restrictions (as a admin not a user). 

The goal is to retrieve the data from a collection in which the search argument is a field in a relation in another collection (with a relation pointing to it) : 
As in my previous example, i search records in articles written by an author (relation pointing to authors collection as a single relation).
Then, after retrieving the data with expanded relations do some processing and then create new records in another collection (as an admin again). I plan to interact with the expanded records in a JSON object. It's easier to interact with a single JSON object that contains a list of articles and each article has an expanded author field which can be also be expanded (lets say that authors collection has a relation pointing to a subscription collection, then we can access the subscriptions of an author from one of its written articles, sorry its a very dumb example it would just be too complicated to explain the whole architecture of my project :) ).

That's why it would have been great if the implementation was as easy to use as in the dart SDK. 

If [Dao.FindRecordsByFilter()](https://github.com/pocketbase/pocketbase/blob/7bb33d4453453731cd01d6d810105518cd405593/daos/record.go#L239) was expanding relations and i could use relations in filter it would save me a lot of time.

An idea would be to have 2 methods RecordsByFilter, one with admin access RecordsByFilterAdmin (for system tasks for example) and one with user access RecordsByFilterUser (to use in custom routes for example, passing the user to check for permissions).

**Reply by ganigeorgiev** [reply, marked as answer]
Thank you for clarifying. Expanding relations and searching in nested relation fields internally are very different things, especially for the relations expansion because often it requires performing some access checks. But this doesn't seem the case in your project.

> I plan to interact with the expanded records in a JSON object. It's easier to interact with a single JSON object that contains a list of articles and each article...

Unfortunately I don't think this would be the case with Go (or any other strongly typed language), unless you map the JSON data in a struct, because if you use something like `map[string]any` you'll need type assertion at almost every access level.

In any case, there will be more detailed db guide with the upcoming v0.17.0 release, but as an example, you can currently do something like:

```go
records := []*models.Record{}

collection, err := app.Dao().FindCollectionByNameOrId("articles")
if err != nil {
	return err
}

// retrieve the articles matching the specified author
queryErr := app.Dao().RecordQuery(collection).
	InnerJoin("authorCollection as author", dbx.HashExp{"author.name": "Jack"}).
	All(&records)
if queryErr != nil {
	return queryErr
}

// eager load the author relation as expand
// I'll consider eventually simplifying the below and allow passing `nil`
// or a predefined DefaultExpandFunc to minimize the verbosity of calling FindRecordsByIds
app.Dao().ExpandRecords(records, []string{"author.otherNestedRelField"}, func(relCollection *models.Collection, relIds []string) ([]*models.Record, error) {
	return app.Dao().FindRecordsByIds(relCollection.Id, relIds)
})

// skipped the type assertion OK checks
// (I'll consider adding SingleExpand(relField) and MultipleExpand(relField) helpers that will normalize the access to the fields)
log.Println(records[0].Expand()["author"].(*models.Record).Expand()["otherNestedRelField"])
```

---

_Just for completeness, with v0.17.0 you'll be able to do the above also with JS (with the added benefit in your case of easier access to the dynamic data):_

```js
const records = arrayOf(new Record)

const collection = $app.dao().findCollectionByNameOrId("articles")

// retrieve the articles matching the specified author
$app.dao().recordQuery(collection)
    .innerJoin("authorCollection as author", $dbx.hashExp({"author.name": "Jack"}))
    .all(records)

// eager load the author relation as expand
$app.dao().expandRecords(records, ["author.otherNestedRelField"], (relCollection, relIds) => {
    return $app.dao().findRecordsByIds(relCollection.id, relIds)
})

console.log(records[0].expand()?.author?.expand()?.otherNestedRelField)
```
